app/lib/annotation_viz_service.rb

```diff
@@ -83,6 +83,24 @@ def self.populate_annotation_by_class(source:, scope:, type:)
     annotation
   end
 
+  # create a hash of cell names => annotations for use in visualization
+  # this is required for all expression-based scenarios to allow filtering of data by observed cells in matrices
+  # without the annotation map, array index position becomes unreliable and annotation assignments will drift
+  def self.get_annotation_as_hash(study, cluster, annotation, subsample_threshold = nil)
+    subsample_annotation = "#{annotation[:name]}--#{annotation[:type]}--#{annotation[:scope]}"
+    case annotation[:scope]
+    when 'study'
+      study.cell_metadata.by_name_and_type(annotation[:name], annotation[:type])&.cell_annotations || {}
+    else
+      # cluster- and user-based annotations behave the same in this case
+      cells = cluster.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
```

Comment:
BLOCKER:  improve performance of the method by taking a (possibly optional) cluster_cells argument to this `get_annotation_as_hash` method. In the places where this method is called, it looks like we've already run the cluster.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation) query, so we can save running it twice

Reply:
That's a great idea - will do.